Upgrade Improvements, Asset Changes & Internal Door Fix

Jeryn99 opened the pull request:
This PR depends on https://github.com/Suff99/TardisRefined/pull/182 being merged 

It currently updates some existing textures upon @Magicmaan request and adds the Lift Exterior.

On the upgrade side of things, it updates the packets used within upgrades to update the screen upon a player purchasing a new upgrade, as previously it did not do this and instead would still show the upgrade in a locked state even though the upgrade was not locked and would often let the user buy it multiple times (Although, without wasting points as the Server packet would deny the request) - due to upgrades affecting the monitor with upgrades such as Desktop Reconfiguration and the Chameleon Circuit, these had to be hidden from the monitor, although this made the monitor look bare and would appear to a user as if it was not functioning correctly. There has now been a message added to the Monitor Screen to indicate that there is no systems available until at least one of these upgrades has been unlocked.

These changes also affected the Dimension control, so a Hardware offline message was assigned to the control and will be displayed to the user via a hot bar message.


Changes were made to the way models are registered, there is now a ShellEntry class that tags the a ShellModel and a ShellDoorModel together, new shells can be registered in the ShellModelCollection class 

This PR is a draft, as I wish to provide more documentation to the classes involved, but a review would be lovely 



## Upgrades
- ARS Tree now grows with the percentage of upgrades unlocked
- External Shell Configuration now locked behind Upgrades
- x1000 Increment now locked behind upgrades
- x2500 Increment now introduced and locked behind upgrades
- x5000 Increment now introduced and locked behind upgrades
- Added Waypoints 
- Added Co-ordinate Input (This might make you lazy people happy, but you are going to hell and back to get it)
- Landing Pad now locked behind upgrades

## Particles
#### Additions
- Added ARS Leaves Particles
- Crashed TARDIS now emits smokey particles as if room is overflowing with smoke
- Root Plant now emits particles when growing to indicate to the user that they are doing it correctly

## Sounds
#### Updates
- Added ARS Room sound
- Added Destination Ding noise, plays when TARDIS initially starts landing
- TARDIS flight noise now loops via LoopingSound

## Commands
#### Additions 
- Introduced Upgrades command to unlock/lock TARDIS Upgrades
- Introduced Points command to set TARDIS xp/points

## Shells
#### Updates
- Updated Police Box Shell Model
- Updated Factory Shell Model
#### Additions
- Added Lift Shell
- Added Hieroglyph
- Added Castle
#### Removals
- Removed Police Box "Marbles" Pattern due to Model Change
- Removed Police Box "Gaudy" Pattern due to Model Change
- Removed Police Box "Metal" Pattern due to Model Change
- Removed Police Box "Stone" Pattern due to Model Change
- Removed Police Box "Red" Pattern due to Model Change

## Crafting
- Bulkhead doors can now be crafted

## Consoles
#### Additions
- Updated all Toyota Pattern Textures
#### Updates
- Updated Nuka Default Pattern

## Interiors
- Added Shalka Interior

## TARDIS Growth Process 
- Players must now use Shears to unlock the growth stage TARDIS

## Rooms
- ARS Room is back under mod protection
- ARS Room Tree grows as the user upgrades their TARDIS (will degrow if addons add new upgrades as it is calculated with upgrade unlock percentage)

## Immersive Portals (Fabric)
- Fixed bug where users could not leave via Immersive Portals portals due to a sizing issue

## Quality of Life
- Interior doors and Shells now hold their own data instead of taking it from their related dimension
- Shell animations now play in Shell Selection Screen
- Sounds that loop are now handled differently, improving performance
- Root shell now only creates the TARDIS dimension when sheared (Hi Performance gains!)
- TARDIS interior is now air instead of growth stone 


# Bugs
- Fix GlobalDoorBlock deleting the player held item when interacted with
- Fix unexpected behaviour with ConsoleConfigurator console removal logic and change theme
- Fixed console removal running the interaction twice and causing the configurator to change its console theme
- Fixed console changing not updating control positions and sizes
- Fixed various screens hard crashing the server

## API
#### Additions
- Added TARDIS Upgrade Unlocked Event
- Allows other mods to add: Shells, Patterns, Desktops, Consoles, Upgrades

50ap5ud5 committed:
Reworked pattern merging

50ap5ud5 committed:
Updated Neoforge for registry changes. Added the ability to support custom registries and holders.

50ap5ud5 committed:
Massive refactor of how Shell and Console Themes and Patterns are handled

50ap5ud5 committed:
Fixed most errors. Waiting on Neoforge network refactor and ForgeConfigAPI for Neoforge

-Made theme translation string be cached to a variable

50ap5ud5 committed:
Account for forge or neoforge non-custom registries

50ap5ud5 committed:
Fix mods.toml still using forge references

50ap5ud5 committed:
Merge branch 'qol/registry_rework' into minecraft/1.20

# Conflicts:
#	common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisExteriorManager.java
#	forge/src/main/java/whocraft/tardis_refined/common/network/forge/NetworkManagerImpl.java

50ap5ud5 committed:
Merge remote-tracking branch 'origin/minecraft/1.20' into minecraft/1.20

# Conflicts:
#	common/src/main/java/whocraft/tardis_refined/common/block/door/GlobalDoorBlock.java
#	forge/build.gradle
#	forge/src/main/resources/META-INF/mods.toml

50ap5ud5 committed:
Made Upgrades use new custom registry design patterns, fixed mods.toml

-Converted upgrades to use our custom registry
-Removed getKey, this is now done by calling the method from the registry itself
-Added a field to cache the translation key instead of relying on the getKey call

50ap5ud5 committed:
Fix DeferredRegistry#createCustom not having matching method signature to its Fabric and Forge implementations

50ap5ud5 committed:
Fix missing registry calls

Jeryn99 committed:
- ARS Egg now opens upgrade screen
- Changed Upgrades UI texture

Jeryn99 committed:
- Locked features behind Upgrades

Jeryn99 committed:
- BOTI Changes
- Sound Changes
- Fixed Waypoint screen

Jeryn99 committed:
- Locked TARDIS summoning behind upgrade

50ap5ud5 committed:
Fixed Configs

Jeryn99 committed:
forge -> neoforge

50ap5ud5 committed:
Re-add overwritten fixes for custom registries not populating registry entries

-Fixed blockstate files still reference the Theme enums
-Fixed lister merger function
-Fixed fabric and forge DeferredRegistry implementations
-Added documentation
-Tweaked wording of CodecJsonReloadListener log message to better reflect how it parses and adds entries.

Jeryn99 committed:
BlockState and Model Data Generation

50ap5ud5 committed:
Fix local merge conflicts

Jeryn99 committed:
Data-Gen, ModelProviderBlock needs looked into, but regardless the patterns have been generated

50ap5ud5 committed:
Fixed pattern sync packets not able to clear map, fixed TardisClientData trying to set a null Theme ID via TardisPilotManager, which doesn't always hold the current Theme ID.

50ap5ud5 committed:
Merge remote-tracking branch 'origin/1.20/registry_rework' into 1.20/registry_rework

# Conflicts:
#	forge/src/generated/.cache/0533c97ce5395f38717098a1ca49d72068853e57
#	forge/src/generated/.cache/07cdbcdc6e4f4553ae9ccc48cd1ff0a5be53a01c
#	forge/src/generated/.cache/103d9f3f36b01595f1aa5172191e60eff02e6924
#	forge/src/generated/.cache/1605092b5c78c792538553a3583c4f114827efd2
#	forge/src/generated/.cache/9fb1092f32d4fcbf9e061ffd718d4ec689c6c95e
#	forge/src/generated/.cache/e7bc600dadd489865d6fbbde8c1c01bd136d8060
#	forge/src/generated/.cache/f681b33583ec37ab445ac361e38402e069dc01f6
#	forge/src/generated/resources/.cache/4dd1d6c274bb02a1ca04c043cc3e81adc6abd533
#	forge/src/generated/resources/.cache/59eb3dbb5f86130e09b3c62d89b9525ee01cf52d
#	forge/src/generated/resources/.cache/735031f3addf80804addae5e3f53249900116f1e
#	forge/src/generated/resources/.cache/c622617f6fabf890a00b9275cd5f643584a8a2c8
#	forge/src/generated/resources/assets/tardis_refined/blockstates/ars_egg.json
#	forge/src/generated/resources/assets/tardis_refined/blockstates/bulk_head_door.json
#	forge/src/generated/resources/assets/tardis_refined/blockstates/internal_door_block.json
#	forge/src/generated/resources/assets/tardis_refined/blockstates/root_plant.json
#	forge/src/generated/resources/assets/tardis_refined/blockstates/root_shell.json
#	forge/src/generated/resources/assets/tardis_refined/blockstates/root_shell_door.json
#	forge/src/generated/resources/assets/tardis_refined/blockstates/tardis_console.json
#	forge/src/generated/resources/assets/tardis_refined/blockstates/tardis_door.json
#	forge/src/generated/resources/assets/tardis_refined/blockstates/tardis_shell.json
#	forge/src/generated/resources/data/minecraft/tags/blocks/fences.json
#	forge/src/generated/resources/data/minecraft/tags/blocks/leaves.json
#	forge/src/generated/resources/data/minecraft/tags/blocks/slabs.json

Jeryn99 committed:
Drastic changes to models

50ap5ud5 committed:
Fixed block entities never saving values to disk when needed, documented and cleaned up Tardis setup process.

-Added much needed documentation about the thought process and purposes of the various setTheme methods. This helps to prevent misuse of the method.
-Made ShellBaseBlockEntity#onBlockInit be a check for if we should generate a Tardis dimension. It now uses a boolean type so that we can stop duplicate dimensions being created if the player clicks on the same root shell twice.
-Fixed Tardis ID in ShellBaseBlockEntity being called the client as well as server during root shell block interaction
-Made The Tardis ID field be a ResourceKey, reduces unnecessary resource key creation for cient side animations which run every tick. Increases performance
-Added cached versions of default fallback patterns, significantly increases client side performance for renderers which lookup patterns every tick.

50ap5ud5 committed:
Fixed Neoforge datagen run configs

50ap5ud5 committed:
Deleted Datagenned files temporarily

50ap5ud5 committed:
Make TardisClientData use a dummy Pattern on game launch, fixes Neoforge crash when TardisClientData is ticked on game load when registries aren't populated yet

Jeryn99 committed:
Drastic changes to models

Jeryn99 committed:
Added message to the Monitor to indicate no upgrades

50ap5ud5 committed:
Fix ConsoleConfigurator not setting correct pattern when changing console, refactored logic to be more clear

-Reduced duplicate code
-Made GlobalConsoleBlockEntity ask for immediate block update when setting important attributes such as theme and pattern

50ap5ud5 committed:
Merge branch '1.20/registry_rework' of https://github.com/Suff99/TardisRefined into 1.20/registry_rework

50ap5ud5 committed:
Fix unexpected behaviour with ConsoleConfigurator console removal logic and change theme

- Fixed console removal running the interaction twice and causing the configurator to change its console theme
- Fixed console changing not updating control positions and sizes

50ap5ud5 committed:
Fixed TardisExteriorManager#setLocked not reflecting its state on the blockstate

50ap5ud5 committed:
Fix ConsoleConfigurationBlock to not cycle themes when placing a console block.

50ap5ud5 committed:
Fix GlobalDoorBlock deleting the player held item when interacted with

Jeryn99 committed:
Lift Shell

50ap5ud5 committed:
Fix GlobalShellBlock setting LOCKED property to true by default

50ap5ud5 committed:
Fixed Tardis Shell not updating REGEN property when interior change has finished

- Caused by erroneous logic in TardisExteriorManager#setShellTheme and TardisInteriorManager#setShellTheme

Jeryn99 committed:
Refactoring to AestheticHandler

Jeryn99 committed:
Changes

Jeryn99 committed:
Indroducing AestheticHandler

Jeryn99 committed:
- Removes Debug Code from Upgrades
- Adds ShellEntry for Addon Mods
- Fixed Upgrades screens not updating when buying a new upgrade
- Refactored Shells Package

Jeryn99 committed:
- Adds Hieroglyph Shell
- Adds Castle Shell

50ap5ud5 commented on common/src/main/java/whocraft/tardis_refined/client/renderer/blockentity/console/GlobalConsoleRenderer.java:
```diff
@@ -47,21 +44,22 @@ public void render(GlobalConsoleBlockEntity blockEntity, float partialTick, Pose
 
     private void renderHoloShell(Vec3 offset, int rotation, GlobalConsoleBlockEntity blockEntity, PoseStack poseStack, MultiBufferSource bufferSource, int packedLight, Vec3 color) {
-        if (blockEntity.getLevel().random.nextInt(20) != 0) {
+       //TODO Implement
```
Needs to use the AestheticsHandler potentially?

Suff99 replied:
Yes, just trying to think of a good solution for one - but I think I have one that's kinda horrible

50ap5ud5 reviewed (changes requested):
Looking good, some things leftover to implement due to the AestheticsHandler refactor.

Jeryn99 committed:
- Trying to fix holoshells, it is not going well,

Jeryn99 committed:
- Upgrades now make the ARS Tree Grow
- ARS room now under protectedzones
- Music, Particles and Fog for ARS Room
- Updated ARS Egg Model
- Fixed bad translation key
- Fixed bug where every block ever would be assigned the wrong render type on Fabric

Jeryn99 committed:
- Upgrades can now be understood
- ARS Eggs shake a lot when crashing now
- Fog color for ARS room goes red ;o

Jeryn99 committed:
- Finalized Upgrades Screen

Jeryn99 committed:
Resolved Merge Conflicts

Jeryn99 committed:
Fixed Holo Shells

Jeryn99 committed:
Shalka Interior

Jeryn99 committed:
ARS Particles, fixed missing toyota textures, fixed bug where shell doesn't update (this was done by @50ap5ud5 before, but somehow got lost in the merge)

Jeryn99 committed:
- Fixed another Toyota pattern
- Eggs in ARS rooms with no Eggs
- Changed Egg model

Jeryn99 committed:
- Updated Factory Shell

Jeryn99 committed:
- Updated Tardis Model (RIP variants)
- Added Tardis Hums (kinda)

Jeryn99 committed:
- Fixed Hum not playing

Jeryn99 committed:
- Commented out redundant Police Box Variants

Magicmaan committed:
Updated Nuka console texture/emission

reworked texture for nuka

Magicmaan committed:
updated factory shell texture

Magicmaan committed:
Update factory_interior.png

Jeryn99 committed:
Upgrade XP/Points command, Dhi better be happy

Jeryn99 committed:
Maven Uploads & Upgrades lock/unlock command

Github Actions will not push a snapshot to our maven on project push, publish.yml will need updated when we merge

Jeryn99 committed:
WIP Changelog

leafreynolds committed:
Prevent Auto Jump making you miss the immersive portal (#187)

(cherry picked from commit fcef0a491fe3ba5da4c78a804ab3dea91dab576e)

leafreynolds committed:
Exposing more functions (#183)

(cherry picked from commit fcadb082acf44d766046fe90a0928e38422a391b)

Jeryn99 committed:
maven publishing is now dates

Jeryn99 committed:
maven publishing is now dates, but more readable

Jeryn99 committed:
Fixed flight_detector.json and landing_pad.json not having the correct render type

(cherry picked from commit a07130befe25768e7e28af5a15fadc39fdfa7536)

Jeryn99 committed:
- Fixed the game breaking itself over Maven dates??
- Bulkhead Doors can no longer be used to nuke walls

Jeryn99 committed:
Allows Addon Makers to register new increment Upgrades

Jeryn99 committed:
Adds Event for allowing/disallowing players the use of controls

Jeryn99 committed:
Fixes Maven for the command line without affecting Github Actions

Jeryn99 committed:
- Fixed Particles renderer not being registered on Forge
- Fixes Controls event acting weirdly

Jeryn99 committed:
Updated publish.yml
